daemon/containerd/image_delete.go

```diff
@@ -59,28 +63,269 @@ func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force,
+
+	if isImageIDPrefix(imgID.String(), imageRef) {
+		return i.deleteAll(ctx, img, force, prune)
+	}
+
+	singleRef, err := i.isSingleReference(ctx, img)
+	if err != nil {
+		return nil, err
+	}
+	if !singleRef {
+		err := i.client.ImageService().Delete(ctx, img.Name)
+		if err != nil {
+			return nil, err
+		}
+		records := []types.ImageDeleteResponseItem{{Untagged: reference.FamiliarString(reference.TagNameOnly(parsedRef))}}
+		return records, nil
+	}
+
+	using := func(c *container.Container) bool {
+		return c.ImageID == imgID
+	}
+	ctr := i.containers.First(using)
+	if ctr != nil {
+		if !force {
+			// If we removed the repository reference then
+			// this image would remain "dangling" and since
+			// we really want to avoid that the client must
+			// explicitly force its removal.
+			refString := reference.FamiliarString(reference.TagNameOnly(parsedRef))
+			err := errors.Errorf(
+				"conflict: unable to remove repository reference %q (must force) - container %s is using its referenced image %s",
+				refString, stringid.TruncateID(ctr.ID), stringid.TruncateID(imgID.String()))
+			return nil, errdefs.Conflict(err)
+		}
+
+		err := i.softImageDelete(ctx, img)
+		if err != nil {
+			return nil, err
+		}
+
+		records := []types.ImageDeleteResponseItem{{Untagged: reference.FamiliarString(reference.TagNameOnly(parsedRef))}}
+		return records, nil
+	}
+
+	return i.deleteAll(ctx, img, force, prune)
+}
+
+// deleteAll deletes the image from the daemon, and if prune is true,
+// also deletes dangling parents if there is no conflict in doing so.
+// Parent images are removed quietly, and if there is any issue/conflict
+// it is logged but does not halt execution/an error is not returned.
+func (i *ImageService) deleteAll(ctx context.Context, img images.Image, force, prune bool) ([]types.ImageDeleteResponseItem, error) {
+	var records []types.ImageDeleteResponseItem
+
 	possiblyDeletedConfigs := map[digest.Digest]struct{}{}
-	if err := i.walkPresentChildren(ctx, img.Target, func(_ context.Context, d ocispec.Descriptor) {
+	err := i.walkPresentChildren(ctx, img.Target, func(_ context.Context, d ocispec.Descriptor) {
 		if images.IsConfigType(d.MediaType) {
 			possiblyDeletedConfigs[d.Digest] = struct{}{}
 		}
-	}); err != nil {
+	})
+	if err != nil {
 		return nil, err
 	}
 
-	err = i.client.ImageService().Delete(ctx, img.Name, images.SynchronousDelete())
+	var parents []imageWithRootfs
+	if prune {
+		parents, err = i.parents(ctx, image.ID(img.Target.Digest))
+		if err != nil {
+			logrus.WithError(err).Warn("failed to get image parents")
+		}
+		sortParentsByAffinity(parents)
+	}
+
+	err = i.imageDeleteHelper(ctx, img, &records, force)
 	if err != nil {
 		return nil, err
 	}
 
+	for _, parent := range parents {
+		if !isDanglingImage(parent.img) {
+			break
+		}
+		err = i.imageDeleteHelper(ctx, parent.img, &records, false)
+		if err != nil {
+			logrus.WithError(err).Warn("failed to remove image parent")
+			break
+		}
```

Comment:
```suggestion
		if err != nil {
			return records, errdefs.System(errors.Wrap(err, "failed to remove image parent"))
		}
```

I understand that logging the image events and `i.unleaseSnapshotsFromDeletedConfigs` can't be short-circuited; those operations can be wrapped up in a `defer` and moved above this loop.